pkg/applications/helmclient/client.go

```diff
@@ -241,10 +246,86 @@ func (h HelmClient) DownloadChart(url string, chartName string, version string,
 // Otherwise it upgrades the chart.
 // charLoc is the path to the chart archive (e.g. /tmp/foo/apache-1.0.0.tgz) or folder containing the chart (e.g. /tmp/mychart/apache).
 func (h HelmClient) InstallOrUpgrade(chartLoc string, releaseName string, values map[string]interface{}, deployOpts DeployOpts, auth AuthSettings) (*release.Release, error) {
-	if _, err := h.actionConfig.Releases.Last(releaseName); err != nil {
+	currentRelease, err := h.actionConfig.Releases.Last(releaseName)
+	if err != nil {
+		h.logger.Debugw("Installing helm release", "release", releaseName)
 		return h.Install(chartLoc, releaseName, values, deployOpts, auth)
 	}
-	return h.Upgrade(chartLoc, releaseName, values, deployOpts, auth)
+
+	upgradeNeeded, err := h.shouldUpgrade(chartLoc, currentRelease, values)
+	if err != nil {
+		return nil, err
+	}
+
+	if upgradeNeeded {
+		h.logger.Debugw("Detected changes between helm releases, running helm upgrade", "release", releaseName)
+		return h.Upgrade(chartLoc, releaseName, values, deployOpts, auth)
+	}
+
+	return currentRelease, nil
+}
+
+func (h HelmClient) shouldUpgrade(chartLoc string, currentRelease *release.Release, values map[string]interface{}) (bool, error) {
+	chart, err := loader.Load(chartLoc)
+	if err != nil {
+		return false, err
+	}
+
+	currentValues, err := chartutil.CoalesceValues(currentRelease.Chart, currentRelease.Config)
+	if err != nil {
+		return false, err
+	}
+
+	newValues, err := chartutil.CoalesceValues(chart, values)
+	if err != nil {
+		return false, err
+	}
+
+	currentManifests, err := h.renderManifests(currentRelease.Chart, currentValues)
+	if err != nil {
+		return false, err
+	}
+
+	newManifests, err := h.renderManifests(chart, newValues)
+	if err != nil {
+		return false, err
+	}
+
+	return !reflect.DeepEqual(chart.AppVersion(), currentRelease.Chart.AppVersion()) ||
```

Comment:
Do we need a `reflect.DeepEqual` for comparing two strings here?

Reply:
ah, I think I left it here, thanks for catching